zookeeper-server/src/main/java/org/apache/zookeeper/ZooKeeper.java

```diff
@@ -2800,26 +2807,100 @@ public void setACL(final String path, List<ACL> acl, int version, StatCallback c
         // the watch contains the un-chroot path
         WatchRegistration wcb = null;
         if (watcher != null) {
-            wcb = new ChildWatchRegistration(watcher, clientPath);
+            wcb = new ChildWatchRegistration(watcher, path);
         }
 
-        final String serverPath = prependChroot(clientPath);
+        final String serverPath = prependChroot(path);
 
         RequestHeader h = new RequestHeader();
         h.setType(ZooDefs.OpCode.getChildrenPaginated);
         GetChildrenPaginatedRequest request = new GetChildrenPaginatedRequest();
         request.setPath(serverPath);
         request.setWatch(watcher != null);
         request.setMaxReturned(maxReturned);
-        request.setMinCzxId(minCzxId);
-        request.setCzxIdOffset(czxIdOffset);
-        GetChildrenPaginatedResponse response = new GetChildrenPaginatedResponse();
-        ReplyHeader r = cnxn.submitRequest(h, request, response, wcb);
-        if (r.getErr() != 0) {
-            throw KeeperException.create(KeeperException.Code.get(r.getErr()),
-                    clientPath);
+
+        Set<String> children = null;
+        GetChildrenPaginatedResponse response;
+        boolean isFirstPage = true;
```

Comment:
I initially did this do..while. Then later I found that in 99% of the cases, we don't need more than one page. Using the `firstPage` variable is to keep the behavior as the non-paginated `getChildren`.

If we use do..while, we'll always add all children to the set and then convert it again to a list. I don't like this redundant cost. And I'd prefer to keep the behavior as the original non-paginated `getChildren` if all the children can be returned in the first page: just return the list `response.getChildren()`, without adding all to set and converting to the final list.

Reply:
Keeping the order or not does not matter at all. This API does not guarantee order.

I still prefer a method which can keep the code clean. The additional cost is minor, and it is on the client side. So I think it would be fine.

Reply:
As discussed, code is updated to split the logic to make the code cleaner.